Expose built-in region information

dsmtE opened the pull request:
Related to this proposal and issue:
- https://github.com/godotengine/godot-proposals/issues/9355
- https://github.com/godotengine/godot/issues/57401

I'm not sure about the defines (`#if !defined(USE_ATTRIBUTES) && !defined(USE_PRIMITIVE)`) but it works on both gl_compatibility and mobile rendering mode.

* *Bugsquad edit, closes: https://github.com/godotengine/godot-proposals/issues/9355*

Mickeon commented:
It may be a good idea to schedule this for a rendering meeting next time.

Delsin-Yu commented:
Hello there, is there any progress to this PR?

jtcaya commented:
Hello, is there any chance that this PR might be included in v4.4?

It would solve a lot of issues with complicated 2D UV-related shaders.

Many thanks!

QbieShay commented:
It won't be in 4.4, no. We're in beta stage and we don't add features in beta stage. I can see this has been open for a while, and I'm sorry for that.

QbieShay commented:
As a workaround for now, you can try using a canvas group and applying your shader to the canvas group.

Delsin-Yu commented:
I guess the team never revisited this PR after Mickeon's comment; they likely forgot. I will raise this in the Rocket chat channel; as QbieShay mentioned, it is too late for 4.4, but we can make it inside the 4.5 dev1.

dsmtE commented:
> It won't be in 4.4, no. We're in beta stage and we don't add features in beta stage. I can see this has been open for a while, and I'm sorry for that. We can tentatively take a look at it for 4.5

Hello @QbieShay,
yes, it's been open for a while. I understand for 4.4, it would be nice if it could be looked at for 4.5.
maybe I can take a little time (in a few weeks because I can't right now in my personal life) to try to rebase the branch and manage the conflicts to be closer to the current code base that has evolved since maybe. So that it's easier to integrate for 4.5?

QbieShay commented:
Don't worry, before you add any code, I think we should have a chat in the rendering meeting about the feature itself and the API, if we want to expose a function like the one you did, or directly give rect info and full UV info (see #32861).

Once this is discussed, we can come back to the implementation, so that you don't code things until we know exactly in which way we want this feature.

jtcaya commented:
Okay, that is understandable.

I will try the canvas group method in the meantime.

Thank you all for your time!

Cheers!

lucastucious commented:
Hey, bumping this topic as 4.5 is starting

KirisameY commented:
Hello, will it be in 4.5?💦

clayjohn reviewed (changes requested):
We discussed this in today's rendering meeting. We agreed that this is a problem with solving, but we think the solution is not the right one. 

Instead we should just be exposing the region info directly to users for a few reasons:
1. Storing those intermediate values in global space wastes a lot of registers which will adversely impact performance
2. Users are unlikely to discover the functions naturally. Instead they will look for ways of figuring out the rect itself
3. This approach adds a lot of API surface for something that doesn't require it. 

Accordingly, I would just expose `REGION_RECT` in the fragment shader and then add corresponding documentation.

dsmtE commented:
Great news! Thank you @clayjohn for taking the time to discuss it and take a look. I propose (I would  be happy to contribute to godot ) to modify my PR to only add and expose REGION_RECT (composed of the position (xy) and size of the region (zw) coming from draw_data.src_rect right?). Or would you prefer to do it yourself?

KirisameY commented:
I think expose it in any case is simpler and eleganter, maybe (0,0,1,1) can be a default value for non-atlas textures

________________________________
寄件者: LUCASTUCIOUS ***@***.***>
寄件日期: 星期三, 4月 2, 2025 12:54:57 上午
收件者: godotengine/godot ***@***.***>
副本: F雨烨 ***@***.***>; Manual ***@***.***>
主旨: Re: [godotengine/godot] Expose built-in region information (PR #90436)

[lucastucious]lucastucious left a comment (godotengine/godot#90436)<https://github.com/godotengine/godot/pull/90436#issuecomment-2769987779>

Will REGION_RECT will be exposed for textures without region? So the same shader can be the used on Atlas and non-atlas

―
Reply to this email directly, view it on GitHub<https://github.com/godotengine/godot/pull/90436#issuecomment-2769987779>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AMPCZ3MO6BC7UNMRVWFMYSL2XLAF3AVCNFSM6AAAAABUOZ5YZOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDONRZHE4DONZXHE>.
You are receiving this because you are subscribed to this thread.

KirisameY commented:
> Will `REGION_RECT` will be exposed for textures without region? So the same shader can be the used on Atlas and non-atlas

I think expose it in any case is simpler and eleganter, maybe (0,0,1,1) can be a default value for non-atlas textures

dsmtE commented:
@QbieShay i will try to take some time to update my PR ;)
Ok for the **visual shaders**, If I have any difficulty doing this part I will come back to you, if you have any entry points to suggest to me in the code I would be interested .

dsmtE committed:
add REGION_RECT

fix formatting

fix shader compile (region_rect after draw_data)

fix formatting

fix formatting

Review fix (move region_rect in main + remove useless visual shader options)

fix formatting

fix alphabetic order (+ remove useless action for spatial shaders)

typo + remove TYPE_LIGHT option for  region_rect

dsmtE commented:
I made the changes and exposed only the REGION_RECT uniform (composed of the position and size of the region used), which allows me to manipulate the region as I like in the shader.

I did a test (shader and visualShader) for the demo and it works great!

Can you give me feedback if there's anything I need to change/refine before I can merge?

In the case where the region is not used, I used the inverse **texture_pixel_size** to set it to the texture size and make the shader work too (it works but I don't know if it's the best way to do it or if there's another way to get the texture size). 

How does the documentation work? I leave it to you?

The shader code used in my test : 
```glsl
shader_type canvas_item;

vec2 zoom_in_region(vec2 uv, float zoom_factor, vec2 region_position, vec2 region_size) {
	// Convert texture space uv to region
	vec2 region_uv = (uv - region_position) / region_size;
	
	// Any local uv transformation : exemple with centered zoom effect
	vec2 zoomed_uv = (region_uv-0.5)/zoom_factor+0.5;
	
	// Convert back to texture uv to be sampled properly using the texture(TEXTURE, uv) function
	return zoomed_uv * region_size + region_position;
}

void fragment() {
	
	vec2 texture_uv = UV;
	
	float time_factor = ((sin(TIME)+1.0)/2.0);
	
	// Apply zoom to the region part using REGION_RECT
	texture_uv = zoom_in_region(texture_uv, 1.0 + 0.4 * time_factor, REGION_RECT.xy, REGION_RECT.zw);
	
	// Shift the region on the x axis using the region size (REGION_RECT.zw) to get the new part of the texture of the same region_size (next sprite sheet frame)
	texture_uv.x += REGION_RECT.z;

    // Sample the texture using right region UV after local transformation
	COLOR = texture(TEXTURE, texture_uv);
}
```
![godotRegionPR](https://github.com/user-attachments/assets/6b3b97bb-c47a-4534-9026-2b9cadccc61d)

- *Upper left : shader*
- *Upper right : visual shader*
- *Bottom : same shader without using region* (texture repeat enable to make the shift work)

dsmtE commented:
Hello,
I finally test it and fix it on all Rendering methods (Forward, Mobile & Compatibility) and it work well.
Could you review my PR @clayjohn @QbieShay  ? Thanks in advance !

clayjohn commented on editor/plugins/visual_shader_editor_plugin.cpp:
```diff
@@ -7101,6 +7103,7 @@ VisualShaderEditor::VisualShaderEditor() {
 	add_options.push_back(AddOption("PointCoord", "Input/Fragment", "VisualShaderNodeInput", vformat(input_param_for_fragment_and_light_shader_modes, "point_coord", "POINT_COORD"), { "point_coord" }, VisualShaderNode::PORT_TYPE_VECTOR_2D, TYPE_FLAGS_FRAGMENT, Shader::MODE_CANVAS_ITEM));
 	add_options.push_back(AddOption("ScreenPixelSize", "Input/Fragment", "VisualShaderNodeInput", vformat(input_param_for_fragment_shader_mode, "screen_pixel_size", "SCREEN_PIXEL_SIZE"), { "screen_pixel_size" }, VisualShaderNode::PORT_TYPE_VECTOR_2D, TYPE_FLAGS_FRAGMENT, Shader::MODE_CANVAS_ITEM));
 	add_options.push_back(AddOption("ScreenUV", "Input/Fragment", "VisualShaderNodeInput", vformat(input_param_for_fragment_and_light_shader_modes, "screen_uv", "SCREEN_UV"), { "screen_uv" }, VisualShaderNode::PORT_TYPE_VECTOR_2D, TYPE_FLAGS_FRAGMENT, Shader::MODE_CANVAS_ITEM));
+	add_options.push_back(AddOption("RegionRect", "Input/Fragment", "VisualShaderNodeInput", vformat(input_param_for_fragment_and_light_shader_modes, "region_rect", "REGION_RECT"), { "region_rect" }, VisualShaderNode::PORT_TYPE_VECTOR_4D, TYPE_FLAGS_FRAGMENT, Shader::MODE_CANVAS_ITEM));
```
Should use `input_param_for_fragment_shader_mode`

clayjohn commented on editor/plugins/visual_shader_editor_plugin.cpp:
```diff
@@ -7060,6 +7060,7 @@ VisualShaderEditor::VisualShaderEditor() {
 	add_options.push_back(AddOption("NodePositionWorld", "Input/Fragment", "VisualShaderNodeInput", vformat(input_param_for_vertex_and_fragment_shader_modes, "node_position_world", "NODE_POSITION_WORLD"), { "node_position_world" }, VisualShaderNode::PORT_TYPE_VECTOR_3D, TYPE_FLAGS_FRAGMENT, Shader::MODE_SPATIAL));
 	add_options.push_back(AddOption("PointCoord", "Input/Fragment", "VisualShaderNodeInput", vformat(input_param_for_fragment_shader_mode, "point_coord", "POINT_COORD"), { "point_coord" }, VisualShaderNode::PORT_TYPE_VECTOR_2D, TYPE_FLAGS_FRAGMENT, Shader::MODE_SPATIAL));
 	add_options.push_back(AddOption("ScreenUV", "Input/Fragment", "VisualShaderNodeInput", vformat(input_param_for_fragment_and_light_shader_modes, "screen_uv", "SCREEN_UV"), { "screen_uv" }, VisualShaderNode::PORT_TYPE_VECTOR_2D, TYPE_FLAGS_FRAGMENT, Shader::MODE_SPATIAL));
+	add_options.push_back(AddOption("RegionRect", "Input/Fragment", "VisualShaderNodeInput", vformat(input_param_for_fragment_and_light_shader_modes, "region_rect", "REGION_RECT"), { "region_rect" }, VisualShaderNode::PORT_TYPE_VECTOR_4D, TYPE_FLAGS_FRAGMENT, Shader::MODE_SPATIAL));
```
Can be removed since the new built in is for canvas items only

clayjohn commented on drivers/gles3/storage/material_storage.cpp:
```diff
@@ -1272,6 +1273,7 @@ MaterialStorage::MaterialStorage() {
 		actions.renames["POINT_COORD"] = "gl_PointCoord";
 		actions.renames["INSTANCE_CUSTOM"] = "instance_custom";
 		actions.renames["SCREEN_UV"] = "screen_uv";
+		actions.renames["REGION_RECT"] = "region_rect";
```
Sorry, missed this before. This can be removed since it is for spatial shaders. Only the definition above is needed

Chaosus commented on editor/plugins/visual_shader_editor_plugin.cpp:
```diff
@@ -7101,6 +7101,7 @@ VisualShaderEditor::VisualShaderEditor() {
 	add_options.push_back(AddOption("PointCoord", "Input/Fragment", "VisualShaderNodeInput", vformat(input_param_for_fragment_and_light_shader_modes, "point_coord", "POINT_COORD"), { "point_coord" }, VisualShaderNode::PORT_TYPE_VECTOR_2D, TYPE_FLAGS_FRAGMENT, Shader::MODE_CANVAS_ITEM));
 	add_options.push_back(AddOption("ScreenPixelSize", "Input/Fragment", "VisualShaderNodeInput", vformat(input_param_for_fragment_shader_mode, "screen_pixel_size", "SCREEN_PIXEL_SIZE"), { "screen_pixel_size" }, VisualShaderNode::PORT_TYPE_VECTOR_2D, TYPE_FLAGS_FRAGMENT, Shader::MODE_CANVAS_ITEM));
 	add_options.push_back(AddOption("ScreenUV", "Input/Fragment", "VisualShaderNodeInput", vformat(input_param_for_fragment_and_light_shader_modes, "screen_uv", "SCREEN_UV"), { "screen_uv" }, VisualShaderNode::PORT_TYPE_VECTOR_2D, TYPE_FLAGS_FRAGMENT, Shader::MODE_CANVAS_ITEM));
+	add_options.push_back(AddOption("RegionRect", "Input/Fragment", "VisualShaderNodeInput", vformat(input_param_for_fragment_shader_mode, "region_rect", "REGION_RECT"), { "region_rect" }, VisualShaderNode::PORT_TYPE_VECTOR_4D, TYPE_FLAGS_FRAGMENT, Shader::MODE_CANVAS_ITEM));
```
It must be declared before `ScreenPixelSize` to achieve alphabetic order.

Chaosus commented on scene/resources/visual_shader.cpp:
```diff
@@ -3203,6 +3203,7 @@ const VisualShaderNodeInput::Port VisualShaderNodeInput::ports[] = {
 	{ Shader::MODE_CANVAS_ITEM, VisualShader::TYPE_FRAGMENT, VisualShaderNode::PORT_TYPE_VECTOR_2D, "point_coord", "POINT_COORD" },
 	{ Shader::MODE_CANVAS_ITEM, VisualShader::TYPE_FRAGMENT, VisualShaderNode::PORT_TYPE_VECTOR_2D, "screen_pixel_size", "SCREEN_PIXEL_SIZE" },
 	{ Shader::MODE_CANVAS_ITEM, VisualShader::TYPE_FRAGMENT, VisualShaderNode::PORT_TYPE_VECTOR_2D, "screen_uv", "SCREEN_UV" },
+	{ Shader::MODE_CANVAS_ITEM, VisualShader::TYPE_FRAGMENT, VisualShaderNode::PORT_TYPE_VECTOR_4D, "region_rect", "REGION_RECT" },
```
Same here

Chaosus commented on scene/resources/visual_shader.cpp:
```diff
@@ -3373,6 +3375,7 @@ const VisualShaderNodeInput::Port VisualShaderNodeInput::preview_ports[] = {
 	{ Shader::MODE_SPATIAL, VisualShader::TYPE_FRAGMENT, VisualShaderNode::PORT_TYPE_VECTOR_4D, "fragcoord", "FRAGCOORD" },
 	{ Shader::MODE_SPATIAL, VisualShader::TYPE_FRAGMENT, VisualShaderNode::PORT_TYPE_VECTOR_3D, "normal", "vec3(0.0, 0.0, 1.0)" },
 	{ Shader::MODE_SPATIAL, VisualShader::TYPE_FRAGMENT, VisualShaderNode::PORT_TYPE_VECTOR_2D, "screen_uv", "SCREEN_UV" },
+	{ Shader::MODE_SPATIAL, VisualShader::TYPE_LIGHT, VisualShaderNode::PORT_TYPE_VECTOR_4D, "region_rect", "REGION_RECT" },
```
typo? Otherwise, it must be declared where similar TYPE_LIGHT entries are declared.

Chaosus commented on scene/resources/visual_shader.cpp:
```diff
@@ -3224,6 +3225,7 @@ const VisualShaderNodeInput::Port VisualShaderNodeInput::ports[] = {
 	{ Shader::MODE_CANVAS_ITEM, VisualShader::TYPE_LIGHT, VisualShaderNode::PORT_TYPE_VECTOR_3D, "normal", "NORMAL" },
 	{ Shader::MODE_CANVAS_ITEM, VisualShader::TYPE_LIGHT, VisualShaderNode::PORT_TYPE_VECTOR_2D, "point_coord", "POINT_COORD" },
 	{ Shader::MODE_CANVAS_ITEM, VisualShader::TYPE_LIGHT, VisualShaderNode::PORT_TYPE_VECTOR_2D, "screen_uv", "SCREEN_UV" },
+	{ Shader::MODE_CANVAS_ITEM, VisualShader::TYPE_LIGHT, VisualShaderNode::PORT_TYPE_VECTOR_4D, "region_rect", "REGION_RECT" },
```
Same here

clayjohn reviewed (approved):
Looks great! The final step before merging is to squash all the commits together so that the whole PR only contains 1 big commit with all your changes. We like to merge one commit at a time to keep the git history clean and navigable. 

If you don't know how to do that, we have a helpful tutorial in the official documentation https://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html#the-interactive-rebase

dsmtE commented:
Thanks for the review @clayjohn  ! Done ! I just rebase my branch on the upstream Master and squash all my commits into one to have a clean history.

oeleo1 commented:
Critical enhancement for serious batching / atlas optimization. Thanks. Can this be backported to 3.x please ?
cc @lawnjelly

oeleo1 commented:
PS: The rationale for my request to backport to 3.x is that batching breaks on shader uniforms. So passing the atlas region rect via uniforms to let the shader do the UV math is a no go (both in Godot 3 and 4). The only way to pass parameters (region rect) to the shader and keep batching is the Color `modulate/self_modulate` of the atlas sprite, and collect that from MODULATE in the shader. But passing the region rect via Color is consuming most of the 4-byte budget for 2048p atlases, leaving no room for other shader instance params, so no go again. Hence, to me this enhancement solves the problem of having the region rect in the fragment shader and is a critical enabler of batchning with atlases. Without it, atlases and batching don't fly.

Thanks to all for solving this, except that our project is Godot 3, but the good news is that it looks like backporting this to GLES2/GLES3 for Godot 3 it totally doable without much effort. cc @lawnjelly

lawnjelly commented:
> Thanks to all for solving this, except that our project is Godot 3, but the good news is that it looks like backporting this to GLES2/GLES3 for Godot 3 it totally doable without much effort.

Unfortunately I suspect this wouldn't be trivial to add to 3.x batching. Adding new parameters to be passed in the FVF is most definitely *not* trivial (this is why it is extremely limited at present). It has to be done so as not to slow down existing paths, then be extensively checked for regressions etc.

I have no plans currently to look at this (it would require significant user demand to be worth it at this stage in Godot 3.x, and speaking for myself I have a lot of other areas on the road map before considering anything like this).

clayjohn commented:
@oeleo1 in 3.x we often bake in the vertex positions on the CPU, so the region rect information is not necessarily always available. In 4.x the region information is always available because it is never baked directly into the vertex data and is always extracted in the shader
